### PR TITLE
add isDeterministic flag to IRFunction to help with optimization

### PR DESCRIPTION
**Pull request by catoverdrive**
the `isDeterministic` flag tells us whether we can expect a fixed output given a fixed input. It is checked when folding constants to preserve randomness.

@jbloom22, I'm giving this to you because I'm not super sure on which math functions need to have this flag turned off.

### CODE REVIEW

**Comment by cseed** [comment]
Want to comment back in the constant foldings?

**Comment by catoverdrive** [comment]
I did that, sorry. Just haven't rebased this branch onto master yet so the commented out lines aren't showing up

File: src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 1]
```diff
@@ -89,8 +89,10 @@ abstract class RegistryFunctions {
   def tnum(name: String): TVariable =
     tv(name, _.isInstanceOf[TNumeric])
 
-  def registerCode(mname: String, aTypes: Array[Type], rType: Type)(impl: (EmitMethodBuilder, Array[Code[_]]) => Code[_]) {
+  def registerCode(mname: String, aTypes: Array[Type], rType: Type, isDet: Boolean)(impl: (EmitMethodBuilder, Array[Code[_]]) => Code[_]) {
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/functions/Functions.scala]: `isDeterministic` more consistent with everywhere else

File: src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 2]
```diff
@@ -101,8 +103,10 @@ abstract class RegistryFunctions {
     })
   }
 
-  def registerCodeWithMissingness(mname: String, aTypes: Array[Type], rType: Type)(impl: (EmitMethodBuilder, Array[EmitTriplet]) => EmitTriplet) {
+  def registerCodeWithMissingness(mname: String, aTypes: Array[Type], rType: Type, isDet: Boolean)(impl: (EmitMethodBuilder, Array[EmitTriplet]) => EmitTriplet) {
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/functions/Functions.scala]: same here

File: src/main/scala/is/hail/expr/ir/functions/Functions.scala (hunk 1 above)
Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/functions/Functions.scala]: I think this causes some weird shadowing to happen when I do 
```
val isDeterministic = isDeterministic
```
inside the IR definition, which is why I named it differently (the second `isDeterministic` tries to pick up the IRFunction.isDeterministic that we're trying to define, and everything ends up false.) I wasn't sure how to fix this problem without renaming the argument.

File: src/main/scala/is/hail/expr/ir/FoldConstants.scala [hunk 3]
```diff
@@ -20,10 +20,19 @@ object FoldConstants {
...
+
         }
     })
 }
+
+object IsDeterministic {
+  def apply(ir: IR): Boolean = ir match {
+    case x: Apply => x.isDeterministic
+    case x: ApplySpecial => x.isDeterministic
+    case _ => true
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/ir/FoldConstants.scala]: this can't be right. Doesn't this have to search recursively?

Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/FoldConstants.scala]: I don't think so? The recursion should still be handled by FoldConstants?

**Pull request merged**